src/freertos_drivers/common/TCAN4550Can.cxx

```diff
@@ -0,0 +1,742 @@
+        Ile ile;
+        register_write(ILE, ile.data);
+    }
+    {
+        // disable MCAN interrupts
+        mcanInterruptEnable_.data = 0;
+        register_write(IE, mcanInterruptEnable_.data);
+    }
+
+    flush_buffers();
+}
+
+//
+// TCAN4550Can::flush_buffers()
+//
+void TCAN4550Can::flush_buffers()
+{
+    // lock SPI bus access
+    OSMutexLock locker(&lock_);
+
+    // cancel TX FIFO buffers
+    register_write(TXBCR, TX_FIFO_BUFFERS_MASK);
+
+    // get the rx status (FIFO fill level)
+    Rxfxs rxf0s;
+    rxf0s.data = register_read(RXF0S);
+
+    /// @todo this could be made more efficeint, but does it matter?
+    while (rxf0s.ffl)
+    {
+        // acknowledge the next FIFO index
+        Rxfxa rxf0a;
+        rxf0a.fai = rxf0s.fgi;
+        register_write(RXF0A, rxf0a.data);
+
+        // increment index and count
+        if (++rxf0s.fgi >= RX_FIFO_SIZE)
+        {
+            rxf0s.fgi = 0;
+        }
+        --rxf0s.ffl;
+    }
+}
+
+//
+// TCAN4550Can::read()
+//
+ssize_t TCAN4550Can::read(File *file, void *buf, size_t count)
+{
+    HASSERT((count % sizeof(struct can_frame)) == 0);
+
+    struct can_frame *data = (struct can_frame*)buf;
+    ssize_t result = 0;
+
+    count /= sizeof(struct can_frame);
+
+    while (count)
+    {
+        size_t frames_read = 0;
+        MRAMRXBuffer *mram_rx_buffer = reinterpret_cast<MRAMRXBuffer*>(data);
+
+        {
+            // lock SPI bus access
+            OSMutexLock locker(&lock_);
+            Rxfxs rxf0s;
+            rxf0s.data = register_read(RXF0S);
+            if (rxf0s.ffl)
+            {
+                static_assert(sizeof(struct can_frame) == sizeof(MRAMRXBuffer));
+
+                // clip to the continous buffer memory available
+                frames_read = std::min(RX_FIFO_SIZE - rxf0s.fgi, rxf0s.ffl);
+
+                // clip to the number of asked for frames
+                frames_read = std::min(frames_read, count);
+
+                // read from MRAM
+                rxbuf_read(
+                    RX_FIFO_0_MRAM_ADDR + (rxf0s.fgi * sizeof(MRAMRXBuffer)),
+                    mram_rx_buffer, frames_read);
+
+                // acknowledge the last FIFO index read
+                Rxfxa rxf0a;
+                rxf0a.fai = rxf0s.fgi + (frames_read - 1);
+                register_write(RXF0A, rxf0a.data);
+            }
+            if (frames_read == rxf0s.ffl)
+            {
+                // all of the data was pulled out, need to re-enable RX
+                // timeout interrupt
+
+                // set pending flag
+                rxPending_ = true;
+
+                // enable receive timeout interrupt
+                mcanInterruptEnable_.too = 1;
+                register_write(IE, mcanInterruptEnable_.data);
+            }
+        }
+        // shuffle data for structure translation
+        for (size_t i = 0; i < frames_read; ++i)
+        {
+            data[i].can_dlc = mram_rx_buffer[i].dlc;
```

Comment:
are these aliased onto each other? I'm wondering if this code is breaking the aliasing rules of the compiler, which would make this code subject to breaking under different compilation settings or a newer version of the compiler.
The only meaningful suggestion i have is to compile the target data in a temporary variable, then add a memory barrier instruction `std::atomic_thread_fence()`, then write it out.